<!DOCTYPE html>
<html>
<head><title>R: Analysis of Massive SNP Arrays</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> Analysis of Massive SNP Arrays
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;bigsnpr&rsquo; version 1.12.15</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
<li><a href="../NEWS">Package NEWS</a>.</li>
</ul>

<h2>Help Pages</h2>


<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="bed-class.html">bed</a></td>
<td>Class bed</td></tr>
<tr><td style="width: 25%;"><a href="bed-class.html">bed-class</a></td>
<td>Class bed</td></tr>
<tr><td style="width: 25%;"><a href="snp_autoSVD.html">bed_autoSVD</a></td>
<td>Truncated SVD while limiting LD</td></tr>
<tr><td style="width: 25%;"><a href="snp_clumping.html">bed_clumping</a></td>
<td>LD clumping</td></tr>
<tr><td style="width: 25%;"><a href="snp_cor.html">bed_cor</a></td>
<td>Correlation matrix</td></tr>
<tr><td style="width: 25%;"><a href="bed_counts.html">bed_counts</a></td>
<td>Counts</td></tr>
<tr><td style="width: 25%;"><a href="bed_cprodVec.html">bed_cprodVec</a></td>
<td>Cross-product with a vector</td></tr>
<tr><td style="width: 25%;"><a href="snp_ld_scores.html">bed_ld_scores</a></td>
<td>LD scores</td></tr>
<tr><td style="width: 25%;"><a href="bed_MAF.html">bed_MAF</a></td>
<td>Allele frequencies</td></tr>
<tr><td style="width: 25%;"><a href="snp_pcadapt.html">bed_pcadapt</a></td>
<td>Outlier detection</td></tr>
<tr><td style="width: 25%;"><a href="bed_prodVec.html">bed_prodVec</a></td>
<td>Product with a vector</td></tr>
<tr><td style="width: 25%;"><a href="bed_projectPCA.html">bed_projectPCA</a></td>
<td>Projecting PCA</td></tr>
<tr><td style="width: 25%;"><a href="bed_projectSelfPCA.html">bed_projectSelfPCA</a></td>
<td>Projecting PCA</td></tr>
<tr><td style="width: 25%;"><a href="bed_randomSVD.html">bed_randomSVD</a></td>
<td>Randomized partial SVD</td></tr>
<tr><td style="width: 25%;"><a href="bed-class.html">bed_RC</a></td>
<td>Class bed</td></tr>
<tr><td style="width: 25%;"><a href="bed_scaleBinom.html">bed_scaleBinom</a></td>
<td>Binomial(2, p) scaling</td></tr>
<tr><td style="width: 25%;"><a href="bed_tcrossprodSelf.html">bed_tcrossprodSelf</a></td>
<td>tcrossprod / GRM</td></tr>
<tr><td style="width: 25%;"><a href="bigSNP-class.html">bigSNP</a></td>
<td>Class bigSNP</td></tr>
<tr><td style="width: 25%;"><a href="bigSNP-class.html">bigSNP-class</a></td>
<td>Class bigSNP</td></tr>
<tr><td style="width: 25%;"><a href="coef_to_liab.html">coef_to_liab</a></td>
<td>Liability scale</td></tr>
<tr><td style="width: 25%;"><a href="download_1000G.html">download_1000G</a></td>
<td>Download 1000G</td></tr>
<tr><td style="width: 25%;"><a href="download_beagle.html">download_beagle</a></td>
<td>Download Beagle 4.1</td></tr>
<tr><td style="width: 25%;"><a href="snp_asGeneticPos2.html">download_genetic_map</a></td>
<td>Download a genetic map</td></tr>
<tr><td style="width: 25%;"><a href="download_plink.html">download_plink</a></td>
<td>Download PLINK</td></tr>
<tr><td style="width: 25%;"><a href="download_plink.html">download_plink2</a></td>
<td>Download PLINK</td></tr>
<tr><td style="width: 25%;"><a href="LD.wiki34.html">LD.wiki34</a></td>
<td>Long-range LD regions</td></tr>
<tr><td style="width: 25%;"><a href="same_ref.html">same_ref</a></td>
<td>Determine reference divergence</td></tr>
<tr><td style="width: 25%;"><a href="SCT.html">SCT</a></td>
<td>Stacked C+T (SCT)</td></tr>
<tr><td style="width: 25%;"><a href="seq_log.html">seq_log</a></td>
<td>Sequence, evenly spaced on a logarithmic scale</td></tr>
<tr><td style="width: 25%;"><a href="snp_ancestry_summary.html">snp_ancestry_summary</a></td>
<td>Estimation of ancestry proportions</td></tr>
<tr><td style="width: 25%;"><a href="snp_asGeneticPos.html">snp_asGeneticPos</a></td>
<td>Interpolate to genetic positions</td></tr>
<tr><td style="width: 25%;"><a href="snp_asGeneticPos2.html">snp_asGeneticPos2</a></td>
<td>Download a genetic map</td></tr>
<tr><td style="width: 25%;"><a href="snp_attach.html">snp_attach</a></td>
<td>Attach a "bigSNP" from backing files</td></tr>
<tr><td style="width: 25%;"><a href="snp_attachExtdata.html">snp_attachExtdata</a></td>
<td>Attach a "bigSNP" for examples and tests</td></tr>
<tr><td style="width: 25%;"><a href="snp_autoSVD.html">snp_autoSVD</a></td>
<td>Truncated SVD while limiting LD</td></tr>
<tr><td style="width: 25%;"><a href="snp_beagleImpute.html">snp_beagleImpute</a></td>
<td>Imputation</td></tr>
<tr><td style="width: 25%;"><a href="snp_clumping.html">snp_clumping</a></td>
<td>LD clumping</td></tr>
<tr><td style="width: 25%;"><a href="snp_cor.html">snp_cor</a></td>
<td>Correlation matrix</td></tr>
<tr><td style="width: 25%;"><a href="snp_fastImpute.html">snp_fastImpute</a></td>
<td>Fast imputation</td></tr>
<tr><td style="width: 25%;"><a href="snp_fastImputeSimple.html">snp_fastImputeSimple</a></td>
<td>Fast imputation</td></tr>
<tr><td style="width: 25%;"><a href="snp_fst.html">snp_fst</a></td>
<td>Fixation index (Fst)</td></tr>
<tr><td style="width: 25%;"><a href="snp_gc.html">snp_gc</a></td>
<td>Genomic Control</td></tr>
<tr><td style="width: 25%;"><a href="snp_getSampleInfos.html">snp_getSampleInfos</a></td>
<td>Get sample information</td></tr>
<tr><td style="width: 25%;"><a href="SCT.html">snp_grid_clumping</a></td>
<td>Stacked C+T (SCT)</td></tr>
<tr><td style="width: 25%;"><a href="SCT.html">snp_grid_PRS</a></td>
<td>Stacked C+T (SCT)</td></tr>
<tr><td style="width: 25%;"><a href="SCT.html">snp_grid_stacking</a></td>
<td>Stacked C+T (SCT)</td></tr>
<tr><td style="width: 25%;"><a href="snp_clumping.html">snp_indLRLDR</a></td>
<td>LD clumping</td></tr>
<tr><td style="width: 25%;"><a href="snp_lassosum2.html">snp_lassosum2</a></td>
<td>lassosum2</td></tr>
<tr><td style="width: 25%;"><a href="LDpred2.html">snp_ldpred2_auto</a></td>
<td>LDpred2</td></tr>
<tr><td style="width: 25%;"><a href="LDpred2.html">snp_ldpred2_grid</a></td>
<td>LDpred2</td></tr>
<tr><td style="width: 25%;"><a href="LDpred2.html">snp_ldpred2_inf</a></td>
<td>LDpred2</td></tr>
<tr><td style="width: 25%;"><a href="snp_ldsc.html">snp_ldsc</a></td>
<td>LD score regression</td></tr>
<tr><td style="width: 25%;"><a href="snp_ldsc.html">snp_ldsc2</a></td>
<td>LD score regression</td></tr>
<tr><td style="width: 25%;"><a href="snp_ldsplit.html">snp_ldsplit</a></td>
<td>Independent LD blocks</td></tr>
<tr><td style="width: 25%;"><a href="snp_ld_scores.html">snp_ld_scores</a></td>
<td>LD scores</td></tr>
<tr><td style="width: 25%;"><a href="snp_MAF.html">snp_MAF</a></td>
<td>MAF</td></tr>
<tr><td style="width: 25%;"><a href="snp_manhattan.html">snp_manhattan</a></td>
<td>Manhattan plot</td></tr>
<tr><td style="width: 25%;"><a href="snp_match.html">snp_match</a></td>
<td>Match alleles</td></tr>
<tr><td style="width: 25%;"><a href="snp_MAX3.html">snp_MAX3</a></td>
<td>MAX3 statistic</td></tr>
<tr><td style="width: 25%;"><a href="snp_modifyBuild.html">snp_modifyBuild</a></td>
<td>Modify genome build</td></tr>
<tr><td style="width: 25%;"><a href="snp_pcadapt.html">snp_pcadapt</a></td>
<td>Outlier detection</td></tr>
<tr><td style="width: 25%;"><a href="snp_plinkIBDQC.html">snp_plinkIBDQC</a></td>
<td>Identity-by-descent</td></tr>
<tr><td style="width: 25%;"><a href="snp_plinkKINGQC.html">snp_plinkKINGQC</a></td>
<td>Relationship-based pruning</td></tr>
<tr><td style="width: 25%;"><a href="snp_plinkQC.html">snp_plinkQC</a></td>
<td>Quality Control</td></tr>
<tr><td style="width: 25%;"><a href="snp_plinkRmSamples.html">snp_plinkRmSamples</a></td>
<td>Remove samples</td></tr>
<tr><td style="width: 25%;"><a href="snp_prodBGEN.html">snp_prodBGEN</a></td>
<td>BGEN matrix product</td></tr>
<tr><td style="width: 25%;"><a href="snp_PRS.html">snp_PRS</a></td>
<td>PRS</td></tr>
<tr><td style="width: 25%;"><a href="snp_clumping.html">snp_pruning</a></td>
<td>LD clumping</td></tr>
<tr><td style="width: 25%;"><a href="snp_qq.html">snp_qq</a></td>
<td>Q-Q plot</td></tr>
<tr><td style="width: 25%;"><a href="snp_readBed.html">snp_readBed</a></td>
<td>Read PLINK files into a "bigSNP"</td></tr>
<tr><td style="width: 25%;"><a href="snp_readBed.html">snp_readBed2</a></td>
<td>Read PLINK files into a "bigSNP"</td></tr>
<tr><td style="width: 25%;"><a href="snp_readBGEN.html">snp_readBGEN</a></td>
<td>Read BGEN files into a "bigSNP"</td></tr>
<tr><td style="width: 25%;"><a href="snp_readBGI.html">snp_readBGI</a></td>
<td>Read variant info from one BGI file</td></tr>
<tr><td style="width: 25%;"><a href="snp_save.html">snp_save</a></td>
<td>Save modifications</td></tr>
<tr><td style="width: 25%;"><a href="snp_scaleBinom.html">snp_scaleAlpha</a></td>
<td>Binomial(n, p) scaling</td></tr>
<tr><td style="width: 25%;"><a href="snp_scaleBinom.html">snp_scaleBinom</a></td>
<td>Binomial(n, p) scaling</td></tr>
<tr><td style="width: 25%;"><a href="snp_simuPheno.html">snp_simuPheno</a></td>
<td>Simulate phenotypes</td></tr>
<tr><td style="width: 25%;"><a href="snp_split.html">snp_split</a></td>
<td>Split-parApply-Combine</td></tr>
<tr><td style="width: 25%;"><a href="snp_subset.html">snp_subset</a></td>
<td>Subset a bigSNP</td></tr>
<tr><td style="width: 25%;"><a href="snp_thr_correct.html">snp_thr_correct</a></td>
<td>Thresholding and correction</td></tr>
<tr><td style="width: 25%;"><a href="snp_writeBed.html">snp_writeBed</a></td>
<td>Write PLINK files from a "bigSNP"</td></tr>
<tr><td style="width: 25%;"><a href="snp_subset.html">subset.bigSNP</a></td>
<td>Subset a bigSNP</td></tr>
<tr><td style="width: 25%;"><a href="sub_bed.html">sub_bed</a></td>
<td>Replace extension '.bed'</td></tr>
</table>
</div></body></html>
