<!DOCTYPE html>
<html>
<head><title>R: Carlo's little helper</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body><div class="container">
<h1> Carlo's little helper
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;binilib&rsquo; version 0.2.0</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<table style="width: 100%;">
<tr><td style="width: 25%;"><a href="any2entrez.html">any2entrez</a></td>
<td>Convert Symbols or Ensembl gene ids to Entrez Ids</td></tr>
<tr><td style="width: 25%;"><a href="any2symbol.html">any2symbol</a></td>
<td>Convert Entrez or Ensembl gene ids to gene symbols</td></tr>
<tr><td style="width: 25%;"><a href="aREA_single.html">aREA_single</a></td>
<td>aREA for a single regulon</td></tr>
<tr><td style="width: 25%;"><a href="basic_rank_norm.html">basic_rank_norm</a></td>
<td>Basic rank normalization of a gene expression matrix</td></tr>
<tr><td style="width: 25%;"><a href="basic_signature.html">basic_signature</a></td>
<td>basic_signature</td></tr>
<tr><td style="width: 25%;"><a href="brks_heatmap.html">brks_heatmap</a></td>
<td>Center heatmap color gradient at zero</td></tr>
<tr><td style="width: 25%;"><a href="collapse_multi.html">collapse_multi</a></td>
<td>Collapsing many-to-one relationships for gene expression matrices on per sample basis</td></tr>
<tr><td style="width: 25%;"><a href="colorScale.html">colorScale</a></td>
<td>colorScale</td></tr>
<tr><td style="width: 25%;"><a href="colVar.html">colVar</a></td>
<td>colVar</td></tr>
<tr><td style="width: 25%;"><a href="DEtransform.html">DEtransform</a></td>
<td>Variance stabilization transformation for RNAseq data</td></tr>
<tr><td style="width: 25%;"><a href="extract_features.html">extract_features</a></td>
<td>Extract rows or columns from a matrix in order of interest</td></tr>
<tr><td style="width: 25%;"><a href="find_var_genes.html">find_var_genes</a></td>
<td>Finding the most variable features of a matrix</td></tr>
<tr><td style="width: 25%;"><a href="geneLengths.html">geneLengths</a></td>
<td>Information on gene lengths. As derived from TxDb.Hsapiens.UCSC.hg19.knownGene and TxDb.Mmusculus.UCSC.mm9.knownGene annotation</td></tr>
<tr><td style="width: 25%;"><a href="geneSets2regulon.html">geneSets2regulon</a></td>
<td>Convert regular (one-tailed) gene sets to regulons</td></tr>
<tr><td style="width: 25%;"><a href="gsea1-class.html">gsea1-class</a></td>
<td>The gsea1 class</td></tr>
<tr><td style="width: 25%;"><a href="gsea1T.html">gsea1T</a></td>
<td>This function performs one-tailed Gene Set Enrichment Analysis as described by Subramanian et al.</td></tr>
<tr><td style="width: 25%;"><a href="gsea2-class.html">gsea2-class</a></td>
<td>The gsea2 class</td></tr>
<tr><td style="width: 25%;"><a href="gsea_null-methods.html">gsea2-method</a></td>
<td>GSEA null model</td></tr>
<tr><td style="width: 25%;"><a href="gsea2ES.html">gsea2ES</a></td>
<td>ES for 2 tail GSEA (Lim et al., 2009) - two character vectors</td></tr>
<tr><td style="width: 25%;"><a href="gsea2ES_regulon.html">gsea2ES_regulon</a></td>
<td>ES for 2 tail GSEA (Lim et al., 2009) - Regulons</td></tr>
<tr><td style="width: 25%;"><a href="gsea2T.html">gsea2T</a></td>
<td>This function performs GSEA of two gene sets for a given signature</td></tr>
<tr><td style="width: 25%;"><a href="gsea_null-methods.html">gsea_null</a></td>
<td>GSEA null model</td></tr>
<tr><td style="width: 25%;"><a href="gsea_null-methods.html">gsea_null,</a></td>
<td>GSEA null model</td></tr>
<tr><td style="width: 25%;"><a href="gsea_null-methods.html">gsea_null-method</a></td>
<td>GSEA null model</td></tr>
<tr><td style="width: 25%;"><a href="gsea_regulon.html">gsea_regulon</a></td>
<td>This function performs GSEA for the negative and positive targets, respectively, of a regulatory gene.</td></tr>
<tr><td style="width: 25%;"><a href="gs_ovlp.html">gs_ovlp</a></td>
<td>Score character vectors of genes and their overlap with certain gene sets</td></tr>
<tr><td style="width: 25%;"><a href="homologyInfo.html">homologyInfo</a></td>
<td>Human and Mouse Homology data Information on 16469 genes for which there is exactly a 1:1 mapping from a mouse gene id to a human gene id.</td></tr>
<tr><td style="width: 25%;"><a href="homologyTable.html">homologyTable</a></td>
<td>Full homology data as provided by Jax Information on a more complete set of 18548 genes for which the Jax Informatics Portal provides mappings between human and mouse genes. This includes human genes for which there are multiple corresponding mouse genes, and vice versa, mouse genes with multiple corresponding human genes.</td></tr>
<tr><td style="width: 25%;"><a href="kaboxplot.html">kaboxplot</a></td>
<td>Plot quantiles per column for a numeric matrix</td></tr>
<tr><td style="width: 25%;"><a href="lapply_pair.html">lapply_pair</a></td>
<td>lapply function to apply functions in a pair-wise manner between all distinct elements of a list</td></tr>
<tr><td style="width: 25%;"><a href="legend_col.html">legend_col</a></td>
<td>Add a custom color gradient to a base R plot</td></tr>
<tr><td style="width: 25%;"><a href="lrcor.html">lrcor</a></td>
<td>For two matched matrices, compute correlation (Spearman) between two related genes with the first coming from matrix A and the second coming from matrix B.</td></tr>
<tr><td style="width: 25%;"><a href="mouse2human.html">mouse2human</a></td>
<td>Convert mouse gene symbols or entrez ids to human counterparts</td></tr>
<tr><td style="width: 25%;"><a href="mypar.html">mypar</a></td>
<td>Quick changes to the plotting environment</td></tr>
<tr><td style="width: 25%;"><a href="order_heatmap.html">order_heatmap</a></td>
<td>Order the columns of a heatmap within the constraints of a factor Generally, a left to right gradient will be achieved where column means will increase from left to right.</td></tr>
<tr><td style="width: 25%;"><a href="p2z.html">p2z</a></td>
<td>Convert p-values from a two tailed test to quantiles of the normal distribution</td></tr>
<tr><td style="width: 25%;"><a href="pick_probes.html">pick_probes</a></td>
<td>Find best probe if there are multiple per gene</td></tr>
<tr><td style="width: 25%;"><a href="plot.gsea1.html">plot.gsea1</a></td>
<td>Plot results from a one-tailed GSEA</td></tr>
<tr><td style="width: 25%;"><a href="plot.gsea2.html">plot.gsea2</a></td>
<td>Plot 2-tailed enrichment results</td></tr>
<tr><td style="width: 25%;"><a href="plot.gsea2regulon.html">plot.gsea2regulon</a></td>
<td>Plot 2-tailed enrichment results</td></tr>
<tr><td style="width: 25%;"><a href="plothm.html">plothm</a></td>
<td>Plot heatmap</td></tr>
<tr><td style="width: 25%;"><a href="plot_fgseaRes.html">plot_fgseaRes</a></td>
<td>Plot one-tailed GSEA on one signature for multiple gene sets</td></tr>
<tr><td style="width: 25%;"><a href="plot_OneGs_MultSig.html">plot_OneGs_MultSig</a></td>
<td>Plot ranks of a given gene set across several gene expression profiles from a numeric matrix.</td></tr>
<tr><td style="width: 25%;"><a href="plot_OneReg_MultSig.html">plot_OneReg_MultSig</a></td>
<td>Plot two tails of one regulon on multiple signatures</td></tr>
<tr><td style="width: 25%;"><a href="purIST.html">purIST</a></td>
<td>PurIST</td></tr>
<tr><td style="width: 25%;"><a href="randomize_eset.html">randomize_eset</a></td>
<td>randomize_eset</td></tr>
<tr><td style="width: 25%;"><a href="randomize_regulon.html">randomize_regulon</a></td>
<td>Generate a randomized network</td></tr>
<tr><td style="width: 25%;"><a href="read_matrix.html">read_matrix</a></td>
<td>Tidyverse way to read in gene expression matrices</td></tr>
<tr><td style="width: 25%;"><a href="reg2tibble.html">reg2tibble</a></td>
<td>transform a regulon object to a data frame / tibble</td></tr>
<tr><td style="width: 25%;"><a href="regulatorNames.html">regulatorNames</a></td>
<td>Information on 6068 regulatory genes used in the Califanoverse See Alvarez, Nature Genetics, 2016 for details</td></tr>
<tr><td style="width: 25%;"><a href="replace_missings.html">replace_missings</a></td>
<td>Replace missing values (typically in a vector or matrix)</td></tr>
<tr><td style="width: 25%;"><a href="rescale01.html">rescale01</a></td>
<td>Rescale a numeric vector to range between 0 and 1</td></tr>
<tr><td style="width: 25%;"><a href="res_msvip.html">res_msvip</a></td>
<td>Extract msviper results</td></tr>
<tr><td style="width: 25%;"><a href="rowTstat.html">rowTstat</a></td>
<td>Calculate row-wise Welchâ€™s t-statistic</td></tr>
<tr><td style="width: 25%;"><a href="rowVar.html">rowVar</a></td>
<td>rowVar</td></tr>
<tr><td style="width: 25%;"><a href="specific_n.html">specific_n</a></td>
<td>For a signature matrix containing conditions/comparisons in its columns and typically gene names in its rows, this function will find the most specific N genes for a given condition. It will do so by computing for each gene the difference between the individual value for in a given condition and the maximum value among the other conditions.</td></tr>
<tr><td style="width: 25%;"><a href="split_violin.html">split_violin</a></td>
<td>Split violin plot</td></tr>
<tr><td style="width: 25%;"><a href="stouffer_z.html">stouffer_z</a></td>
<td>Stouffer integration of Z scores</td></tr>
<tr><td style="width: 25%;"><a href="topN_mat.html">topN_mat</a></td>
<td>For a numeric matrix containing samples or conditions in its columns and typically genes in its rows, this function will find the top N genes for a given condition. You can choose to retrieve both genes with highest and lowest values, respectively or either one of them.</td></tr>
<tr><td style="width: 25%;"><a href="tpm.html">tpm</a></td>
<td>tpm</td></tr>
<tr><td style="width: 25%;"><a href="vpres2regulon.html">vpres2regulon</a></td>
<td>Extract the top up and down regulated MR per sample</td></tr>
<tr><td style="width: 25%;"><a href="vpsig_limma.html">vpsig_limma</a></td>
<td>viperSignature and nullmodel using limma</td></tr>
<tr><td style="width: 25%;"><a href="write_matrix.html">write_matrix</a></td>
<td>Convenience function for writing gene expression to file</td></tr>
</table>
</div></body></html>
